pyresample/bucket/__init__.py

```diff
@@ -0,0 +1,153 @@
+
+    # Calculate array indices
+    x_idxs = ((proj_x - np.min(x_vect)) / x_res).astype(np.int)
+    y_idxs = ((np.max(y_vect) - proj_y) / y_res).astype(np.int)
+
+    # Get valid index locations
+    idxs = ((x_idxs >= 0) & (x_idxs < adef.width) &
+            (y_idxs >= 0) & (y_idxs < adef.height))
+    y_idxs = da.where(idxs, y_idxs, -1)
+    x_idxs = da.where(idxs, x_idxs, -1)
+
+    return x_idxs, y_idxs
+
+
+def get_sum_from_bucket_indices(data, x_idxs, y_idxs, target_shape):
+    """Resample the given data with drop-in-a-bucket resampling.  Return
+    the sum and counts for each bin as two Numpy arrays."""
+
+    if isinstance(data, xr.DataArray):
+        data = data.data
+    data = data.ravel()
+    # Remove NaN values from the data when used as weights
+    weights = da.where(np.isnan(data), 0, data)
+
+    # Convert X- and Y-indices to raveled indexing
+    idxs = y_idxs * target_shape[1] + x_idxs
+    # idxs = idxs.ravel()
+
+    out_size = target_shape[0] * target_shape[1]
+
+    # Calculate the sum of the data falling to each bin
+    sums, _ = da.histogram(idxs, bins=out_size, range=(0, out_size),
+                           weights=weights, density=False)
+    return sums.reshape(target_shape)
+
+
+def get_count_from_bucket_indices(x_idxs, y_idxs, target_shape):
+    """Resample the given data with drop-in-a-bucket resampling.  Return
+    the sum and counts for each bin as two Numpy arrays."""
+
+    # Convert X- and Y-indices to raveled index
+    idxs = y_idxs * target_shape[1] + x_idxs
+
+    out_size = target_shape[0] * target_shape[1]
+
+    # Calculate the sum of the data falling to each bin
+    counts, _ = da.histogram(idxs, bins=out_size, range=(0, out_size))
+
+    return counts.reshape(target_shape)
+
+
+def resample_bucket_average(adef, data, lons, lats,
+                            fill_value=np.nan, x_idxs=None, y_idxs=None):
+    """Calculate bin-averages using bucket resampling."""
+    if x_idxs is None or y_idxs is None:
+        x_idxs, y_idxs = get_bucket_indices(adef, lons, lats)
+    sums = get_sum_from_bucket_indices(data, x_idxs, y_idxs, adef.shape)
+    counts = get_count_from_bucket_indices(x_idxs, y_idxs, adef.shape)
+
+    average = sums / counts
+    average = da.where(counts == 0, fill_value, average)
+
+    return average
+
+
+def resample_bucket_fractions(adef, data, lons, lats, categories,
+                              fill_value=np.nan, x_idxs=None, y_idxs=None):
+    """Get fraction of occurences for each given categorical value."""
+    if x_idxs is None or y_idxs is None:
+        x_idxs, y_idxs = get_bucket_indices(adef, lons, lats)
+    results = {}
+    counts = get_count_from_bucket_indices(x_idxs, y_idxs, adef.shape)
+    counts = counts.astype(float)
+    for cat in categories:
+        cat_data = da.where(data == cat, 1.0, 0.0)
+
+        sums = get_sum_from_bucket_indices(cat_data, x_idxs, y_idxs, adef.shape)
+        results[cat] = sums.astype(float) / counts
+
```

Comment:
Fill results with fill value where count is zero?

Reply:
We discussed this here, and thought that we should leave them as `nan`. But then the `fill_value` kwarg should be removed. I think it's better to allow the user to use a fill value if so wishes, so I'll add that.

Reply:
:+1: